pkg/server/secrets-encrypt.go

```diff
@@ -0,0 +1,509 @@
+		return err
+	}
+
+	curKeys, err := getEncryptionKeys(server)
+	if err != nil {
+		return err
+	}
+
+	if err := AppendNewEncryptionKey(&curKeys); err != nil {
+		return err
+	}
+	logrus.Infoln("Adding secrets-encryption key: ", curKeys[len(curKeys)-1])
+
+	if err := writeEncryptionConfig(server, curKeys, true); err != nil {
+		return err
+	}
+	if err := writeEncryptionState(server, EncryptionPrepare, curKeys[0]); err != nil {
+		return err
+	}
+	if err := writeEncryptionHashAnnotation(server, server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	return cluster.Save(ctx, server, server.Runtime.EtcdConfig, true)
+}
+
+func encryptionRotate(ctx context.Context, server *config.Control, force bool) error {
+	state, err := getEncryptionState(server)
+	if err != nil {
+		return err
+	} else if !force && state.Stage != EncryptionPrepare {
+		return fmt.Errorf("error, incorrect stage: %s found with key: %s", state.Stage, state.ActiveKey)
+	}
+
+	if err := verifyEncryptionHashAnnotation(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+
+	curKeys, err := getEncryptionKeys(server)
+	if err != nil {
+		return err
+	}
+
+	// Right rotate elements
+	rotatedKeys := append(curKeys[len(curKeys)-1:], curKeys[:len(curKeys)-1]...)
+
+	if err = writeEncryptionConfig(server, rotatedKeys, true); err != nil {
+		return err
+	}
+	if err := writeEncryptionState(server, EncryptionRotate, curKeys[0]); err != nil {
+		return err
+	}
+	logrus.Infoln("Encryption keys right rotated")
+	if err := writeEncryptionHashAnnotation(server, server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	return cluster.Save(ctx, server, server.Runtime.EtcdConfig, true)
+}
+
+func encryptionReencrypt(ctx context.Context, server *config.Control, force bool, skip bool) error {
+
+	state, err := getEncryptionState(server)
+	if err != nil {
+		return err
+	} else if !force && state.Stage != EncryptionRotate {
+		return fmt.Errorf("error, incorrect stage: %s found with key: %s", state.Stage, state.ActiveKey)
+	}
+	if err := verifyEncryptionHashAnnotation(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+
+	if err := updateSecrets(server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	if skip {
+		return nil
+	}
+
+	// Remove last key
+	curKeys, err := getEncryptionKeys(server)
+	if err != nil {
+		return err
+	}
+
+	curKeys = curKeys[:len(curKeys)-1]
+	if err = writeEncryptionConfig(server, curKeys, true); err != nil {
+		return err
+	}
+
+	if err := writeEncryptionState(server, EncryptionReencrypt, curKeys[0]); err != nil {
+		return err
+	}
+	logrus.Infoln("Removed key: ", curKeys[len(curKeys)-1])
+	if err := writeEncryptionHashAnnotation(server, server.Runtime.Core.Core()); err != nil {
+		return err
+	}
+	return cluster.Save(ctx, server, server.Runtime.EtcdConfig, true)
+}
+
+func updateSecrets(core core.Interface) error {
+	secrets, err := core.V1().Secret().List("", metav1.ListOptions{})
```

Comment:
I think you're going to need to use pagination here; on relatively large clusters the List call will do bad things when it tries to generate a single Items list containing every secret in the cluster.

https://github.com/kubernetes/kubernetes/pull/75849

Reply:
Moved to pagination within the controller